CRM/Extension/Upgrades.php

```diff
@@ -141,27 +144,19 @@ protected static function sortKeys($keys) {
     $todoKeys = array_unique($keys);
     sort($todoKeys);
 
-    // Goal: Add all active items to $sorter and flag $doneKeys['org.example.foobar']=1.
-    $doneKeys = [];
     $sorter = new \MJS\TopSort\Implementations\FixedArraySort();
 
-    while (!empty($todoKeys)) {
-      $key = array_shift($todoKeys);
-      if (isset($doneKeys[$key])) {
-        continue;
-      }
-      $doneKeys[$key] = 1;
-
+    foreach ($todoKeys as $key) {
       /** @var CRM_Extension_Info $info */
-      $info = @$infos[$key];
+      $info = $infos[$key] ?? NULL;
 
-      if ($info && $info->requires) {
-        $sorter->add($key, $info->requires);
-        $todoKeys = array_merge($todoKeys, $info->requires);
-      }
-      else {
-        $sorter->add($key, []);
+      // Add dependencies
+      if ($info) {
+        // Filter out missing dependencies; missing modules cannot be upgraded
+        $requires = array_intersect($info->requires ?? [], $keys);
+        $sorter->add($key, $requires);
       }
+      // else { what if the extension has no info file? Is that even possible? }
```

Comment:
>  else { what if the extension has no info file? Is that even possible? }

Here are some edge-cases that toy with the existence of `info.xml` and `$info`:

* An extension `child` requires extension `parent`, but `parent` is not available in the source-tree.
    * I believe this is relevant to your intended range of scenarios. Granted, it isn't specifically an issue for `afform_admin` depending on `search_kit` because they're distributed together - so you know the code is there. But that's not true for extensions generally. Suppose  `civivolunteer` added a dependency on `dataprocessor` -- it's quite likely that some (if not all?) users would get the new version of `civivolunteer` first and then encounter errors because of the missing `dataprocessor`. That would prompt them to fix the situation by adding `dataprocessor`.
* An extension `foobar` exists in the source-tree and database, but the developer is currently making edits, and `info.xml` is currently malformed.
    * As above, there could be references to the extension (*eg as a requirement*), but it won't be returned in the `getAllInfos()` list.
* An extension `snafu` exists in the database but was deleted from the filesystem.
    * This one is a little tangential. Note that `CRM_Extension_*` provides a stub `$info` for this -- which ensures that the sysadmin can browse to "Manage Extensions", see/discover the missing item, and restore or disable it. However, from the POV of `CRM_Extension_Upgrades::sortKeys()`, the stub `$info` is already generated. So it's a bit by-the-by.

Reply:
@totten I'm not sure exactly how your comment relates to the question "should this PR be merged" but you don't seem to be saying "don't merge" or flagging any problems with the changes here, so I'll go ahead and add the `merge-ready` tag.

Reply:
Yeah, I agree that it does what it advertises; that it's useful mitigation for search-kit/afform-admin issue.

I think the comment added above can be clarified to indicate an actual use-case (eg `$info` may be missing because the dependency is declared but not present).

Reply:
```suggestion
      // else: We don't have `$info` for some requested extension.
      //       Ex: Existing extension added a `<require>foo</require>` but `foo` is missing.
```

Reply:
Actually I don't think that's a realistic possibility because as of this PR stuff in the `<requires>` tag doesn't get added to the main loop, it only gets added to the topsort.

Reply:
Got, I see how that changed. 👍